src/centrality.rs

```diff
@@ -0,0 +1,296 @@
+        let v = Q.remove(0);
+        verts_sorted_by_distance.push(v);
+        let distance_v = distance[&v];
+        for w in graph.neighbors(v) {
+            if distance[&w] < 0 {
+                Q.push(w);
+                distance.insert(w, distance_v + 1);
+            }
+            if distance[&w] == distance_v + 1 {
+                sigma.insert(
+                    graph.to_index(w),
+                    sigma[&graph.to_index(w)] + sigma[&graph.to_index(v)],
+                );
+                let e_p = predecessors.get_mut(&w).unwrap();
+                e_p.push(v);
+            }
+        }
+    }
+    verts_sorted_by_distance.reverse(); // will be effectively popping from the stack
+    ShortestPathData {
+        verts_sorted_by_distance,
+        predecessors,
+        sigma,
+    }
+}
+
+/// Compute the betweenness centrality of all nodes in a PyGraph.
+///
+/// :param PyGraph graph: The input graph
+/// :param bool normalized: Whether to normalize the betweenness scores by the number of distinct
+///    paths between all pairs of nodes.
+/// :param bool endpoints: Whether to include the endpoints of paths in pathlengths used to
+///    compute the betweenness.
+///
+/// :returns: a dict whose keys are the node indices and values are the betweenness score
+///     for each node.
+/// :rtype: dict
+#[pyfunction(normalized = "true", endpoints = "false")]
+#[pyo3(text_signature = "(graph, /, normalized=True, endpoints=False)")]
+pub fn graph_betweenness_centrality(
+    _py: Python,
+    graph: &graph::PyGraph,
+    normalized: bool,
+    endpoints: bool,
+) -> PyResult<HashMap<usize, f64>> {
+    let betweenness = betweenness_centrality(&graph, endpoints, normalized);
+    let out_map: HashMap<usize, f64> = betweenness
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, v)| v.map(|x| (i, x)))
+        .collect();
+    Ok(out_map)
+}
+
+/// Compute the betweenness centrality of all nodes in a PyDiGraph.
+///
+/// :param PyDiGraph graph: The input graph
+/// :param bool normalized: Whether to normalize the betweenness scores by the number of distinct
+///    paths between all pairs of nodes.
+/// :param bool endpoints: Whether to include the endpoints of paths in pathlengths used to
+///    compute the betweenness.
+///
+/// :returns: a dict whose keys are the node indices and values are the betweenness score
+///     for each node.
+/// :rtype: dict
+#[pyfunction(normalized = "true", endpoints = "false")]
+#[pyo3(text_signature = "(graph, /, normalized=True, endpoints=False)")]
+pub fn digraph_betweenness_centrality(
+    _py: Python,
+    graph: &digraph::PyDiGraph,
+    normalized: bool,
+    endpoints: bool,
+) -> PyResult<HashMap<usize, f64>> {
+    let betweenness = betweenness_centrality(&graph, endpoints, normalized);
+    let out_map: HashMap<usize, f64> = betweenness
+        .into_iter()
+        .enumerate()
+        .filter_map(|(i, v)| v.map(|x| (i, x)))
+        .collect();
+    Ok(out_map)
```

Comment:
I'm wondering if it makes sense to use a custom return type for this. This would avoid the python dict construction cost from the function and return a read-only dict like mapping instead. Off the shelf `PathLengthMapping` wraps a `HashMap<usize, f64>` but the name doesn't make any sense here, but you could add an equivalent new return type by just copying the macro call in `iterators.rs` for that and give it a better name.

Reply:
Done